kratos/processes/vtk_output.cpp

```diff
@@ -0,0 +1,430 @@
+{
+    // write nodal results header
+    Parameters nodal_results = this->mrOutputSettings["nodal_solution_step_data_variables"];
+    rFileStream << "POINT_DATA " << rModelPart.NumberOfNodes() << "\n";
+
+    for (unsigned int entry = 0; entry < nodal_results.size(); entry++)
+    {
+        // write nodal results variable header
+        std::string nodal_result_name = nodal_results[entry].GetString();
+         VtkOutput::WriteDataType dataCharacteristic; // 0: unknown, 1: Scalar value, 2: 3 component vector
+        if (KratosComponents<Variable<double>>::Has(nodal_result_name))
+        {
+            dataCharacteristic =  VtkOutput::WriteDataType::VTK_SCALAR;
+            rFileStream << "SCALARS " << nodal_result_name << " double"
+                       << " 1"
+                       << "\n";
+            rFileStream << "LOOKUP_TABLE default"
+                       << "\n";
+        }
+        else if (KratosComponents<Variable<array_1d<double, 3>>>::Has(nodal_result_name))
+        {
+            dataCharacteristic =  VtkOutput::WriteDataType::VTK_VECTOR;
+            rFileStream << "VECTORS " << nodal_result_name << " double"
+                       << "\n";
+        }
+
+        // write nodal results
+        for(const auto& node : rModelPart.Nodes())
+        {
+            if (dataCharacteristic == VtkOutput::WriteDataType::VTK_SCALAR)
+            {
+                Variable<double> nodal_result_variable = KratosComponents<Variable<double>>::Get(nodal_result_name);
+                const auto &nodal_result = node.FastGetSolutionStepValue(nodal_result_variable);
+                WriteScalarDataToFile(nodal_result, rFileStream);
+                rFileStream <<"\n";
+            }
+            else if (dataCharacteristic == VtkOutput::WriteDataType::VTK_VECTOR)
+            {
+                Variable<array_1d<double, 3>> nodal_result_variable = KratosComponents<Variable<array_1d<double, 3>>>::Get(nodal_result_name);
+                const auto &nodal_result = node.FastGetSolutionStepValue(nodal_result_variable);
+                WriteVectorDataToFile(nodal_result, rFileStream);
+                rFileStream << "\n";
+            }
+        }
+    }
+
+
+}
+
+void VtkOutput::WriteElementData(const ModelPart &rModelPart, std::ofstream& rFileStream)
+{
+    Parameters element_results = this->mrOutputSettings["element_data_value_variables"];// list of element results
+
+    // write cells header
+    if (rModelPart.NumberOfElements() > 0)
+    {
+        rFileStream << "CELL_DATA " << rModelPart.NumberOfElements() << "\n";
+        for (unsigned int entry = 0; entry < element_results.size(); entry++)
+        {
+
+            std::string element_result_name = element_results[entry].GetString();
+            VtkOutput::WriteDataType dataCharacteristic; // 0: unknown, 1: Scalar value, 2: 3 component vector
+
+            if (KratosComponents<Variable<double>>::Has(element_result_name))
```

Comment:
put outside of loop

Reply:
cannot be put outside the loop this need the `element_result_name` which depends on the loop. Also this loop is only on the variables requested to be written, So probably it will be very small number.